lib/constrainer.js

```diff
@@ -38,19 +45,30 @@ class Constrainer {
       throw new Error(`There already exists a custom constraint with the name ${strategy.name}.`)
     }
 
-    if (this.strategiesInUse.has(strategy.name)) {
+    if (this.isStrategyUsed(strategy.name)) {
       throw new Error(`There already exists a route with ${strategy.name} constraint.`)
     }
 
     strategy.isCustom = true
+    strategy.isAsync = strategy.deriveConstraint.length === 3
     this.strategies[strategy.name] = strategy
 
     if (strategy.mustMatchWhenDerived) {
       this.noteUsage({ [strategy.name]: strategy })
     }
   }
 
-  deriveConstraints (req, ctx) {
+  deriveConstraints (req, ctx, done) {
+    const constraints = this.deriveSyncConstraints(req, ctx)
+
+    if (done === undefined) {
+      return constraints
```

Comment:
Sync contraints have priority over the async one?

As example:
https://github.com/Eomm/header-constraint-strategy#cases-explanation

> Multiple constraint matches

How do we want to deal the bullet point 7?

Reply:
I'm not sure that I understand you. You can have a mix of sync and async constraints.
https://github.com/delvedor/find-my-way/blob/2f233391cff8fdd6bdac57c737ea51ed015f8548/test/constraint.custom.async.test.js#L70

The `return` you ask for means that if you use sync deriveConstraint (sync lookup), you will get only sync constraints.

Reply:
I mean this use case:

```
FindMyWay(
{ constraints: { 
   bar: asyncCustomResponseTypeStrategy,
   foo: oldCustomSyncStrategy
 } 
});
```

Reply:
Yes, this will work.

Reply:
It's correct behavior. With the `async: 'bar'` you didn't match the constraint, fallback to the defaultRouter, and you get this error because you pass `null` as a response.

Can you describe what is wrong and how do you want this to work?

Reply:
Sorry, I oversee the issue

I tested a bit more but as you said all constraints are running